// harmonium/lib/Breadcrumbs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BreadcrumbsItem = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BOOL_PROPS_TO_CLASS_NAMES = {
  center: 'Breadcrumbs--center',
  justified: 'Breadcrumbs--justified',
  right: 'Breadcrumbs--right'
};
var BOOL_PROPS = Object.keys(BOOL_PROPS_TO_CLASS_NAMES);

var Breadcrumbs = function (_Component) {
  _inherits(Breadcrumbs, _Component);

  function Breadcrumbs() {
    _classCallCheck(this, Breadcrumbs);

    return _possibleConstructorReturn(this, (Breadcrumbs.__proto__ || Object.getPrototypeOf(Breadcrumbs)).apply(this, arguments));
  }

  _createClass(Breadcrumbs, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['className', 'children']);

      var propClassNames = [];

      BOOL_PROPS.forEach(function (name) {
        if (props[name]) {
          propClassNames.push(BOOL_PROPS_TO_CLASS_NAMES[name]);
        }
        delete props[name];
      });
      var newClassName = (0, _classnames2.default)(className, 'Breadcrumbs', propClassNames);

      return _react2.default.createElement(
        'ul',
        _extends({}, props, { className: newClassName }),
        children
      );
    }
  }]);

  return Breadcrumbs;
}(_react.Component);

Breadcrumbs.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};
exports.default = Breadcrumbs;

var BreadcrumbsItem = exports.BreadcrumbsItem = function (_Component2) {
  _inherits(BreadcrumbsItem, _Component2);

  function BreadcrumbsItem() {
    _classCallCheck(this, BreadcrumbsItem);

    return _possibleConstructorReturn(this, (BreadcrumbsItem.__proto__ || Object.getPrototypeOf(BreadcrumbsItem)).apply(this, arguments));
  }

  _createClass(BreadcrumbsItem, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          className = _props2.className,
          children = _props2.children,
          selected = _props2.selected,
          disabled = _props2.disabled,
          props = _objectWithoutProperties(_props2, ['className', 'children', 'selected', 'disabled']);

      var newClassName = (0, _classnames2.default)('Breadcrumbs-item ' + (className || ''), {
        'Breadcrumbs-item--selected': selected,
        'Breadcrumbs-item--disabled': disabled
      });

      return _react2.default.createElement(
        'li',
        _extends({}, props, { className: newClassName }),
        children
      );
    }
  }]);

  return BreadcrumbsItem;
}(_react.Component);

BreadcrumbsItem.propTypes = {
  selected: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

Breadcrumbs.Item = BreadcrumbsItem;
