// harmonium/lib/Col.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BOOL_PROPS_TO_CLASS_NAMES = {
  // props for float grid
  smallCentered: ['rev-Col--smallCentered'],
  mediumCentered: ['rev-Col--mediumCentered'],
  largeCentered: ['rev-Col--largeCentered'],
  xlargeCentered: ['rev-Col--xlargeCentered'],
  xxlargeCentered: ['rev-Col--xxlargeCentered'],

  smallUncentered: ['rev-Col--smallUncentered'],
  mediumUncentered: ['rev-Col--mediumUncentered'],
  largeUncentered: ['rev-Col--largeUncentered'],
  xlargeUncentered: ['rev-Col--xlargeUncentered'],
  xxlargeUncentered: ['rev-Col--xxlargeUncentered'],

  end: ['rev-Col--end'],

  // padding props for float or flex grid
  collapse: ['rev-Col--collapse'],
  mediumCollapse: ['rev-Col--mediumCollapse'],
  largeCollapse: ['rev-Col--largeCollapse'],
  xlargeCollapse: ['rev-Col--xlargeCollapse'],
  xxlargeCollapse: ['rev-Col--xxlargeCollapse'],

  uncollapse: ['rev-Col--uncollapse'],
  mediumUncollapse: ['rev-Col--mediumUncollapse'],
  largeUncollapse: ['rev-Col--largeUncollapse'],
  xlargeUncollapse: ['rev-Col--xlargeUncollapse'],
  xxlargeUncollapse: ['rev-Col--xxlargeUncollapse'],

  // width props for flex grid
  shrink: ['rev-Col--shrink'],

  // align-self props for flex grid
  alignStart: ['rev-Col--alignStart'],
  alignEnd: ['rev-Col--alignEnd'],
  alignCenter: ['rev-Col--alignCenter'],
  alignBaseline: ['rev-Col--alignBaseline'],
  alignStretch: ['rev-Col--alignStretch']
};

var BOOL_PROPS = Object.keys(BOOL_PROPS_TO_CLASS_NAMES);

var NUMBER_PROPS_TO_CLASS_NAMES = {
  small: function small(arg) {
    return ['rev-Col--small' + arg];
  },
  medium: function medium(arg) {
    return ['rev-Col--medium' + arg];
  },
  large: function large(arg) {
    return ['rev-Col--large' + arg];
  },
  xlarge: function xlarge(arg) {
    return ['rev-Col--xlarge' + arg];
  },
  xxlarge: function xxlarge(arg) {
    return ['rev-Col--xxlarge' + arg];
  },

  smallOffset: function smallOffset(arg) {
    return ['rev-Col--smallOffset' + arg];
  },
  mediumOffset: function mediumOffset(arg) {
    return ['rev-Col--mediumOffset' + arg];
  },
  largeOffset: function largeOffset(arg) {
    return ['rev-Col--largeOffset' + arg];
  },
  xlargeOffset: function xlargeOffset(arg) {
    return ['rev-Col--xlargeOffset' + arg];
  },
  xxlargeOffset: function xxlargeOffset(arg) {
    return ['rev-Col--xxlargeOffset' + arg];
  },

  smallPush: function smallPush(arg) {
    return ['rev-Col--smallPush' + arg];
  },
  mediumPush: function mediumPush(arg) {
    return ['rev-Col--mediumPush' + arg];
  },
  largePush: function largePush(arg) {
    return ['rev-Col--largePush' + arg];
  },
  xlargePush: function xlargePush(arg) {
    return ['rev-Col--xlargePush' + arg];
  },
  xxlargePush: function xxlargePush(arg) {
    return ['rev-Col--xxlargePush' + arg];
  },

  smallPull: function smallPull(arg) {
    return ['rev-Col--smallPull' + arg];
  },
  mediumPull: function mediumPull(arg) {
    return ['rev-Col--mediumPull' + arg];
  },
  largePull: function largePull(arg) {
    return ['rev-Col--largePull' + arg];
  },
  xlargePull: function xlargePull(arg) {
    return ['rev-Col--xlargePull' + arg];
  },
  xxlargePull: function xxlargePull(arg) {
    return ['rev-Col--xxlargePull' + arg];
  },

  smallOrder: function smallOrder(arg) {
    return ['rev-Col--smallOrder' + arg];
  },
  mediumOrder: function mediumOrder(arg) {
    return ['rev-Col--mediumOrder' + arg];
  },
  largeOrder: function largeOrder(arg) {
    return ['rev-Col--largeOrder' + arg];
  },
  xlargeOrder: function xlargeOrder(arg) {
    return ['rev-Col--xlargeOrder' + arg];
  },
  xxlargeOrder: function xxlargeOrder(arg) {
    return ['rev-Col--xxlargeOrder' + arg];
  }
};

var NUMBER_PROPS = Object.keys(NUMBER_PROPS_TO_CLASS_NAMES);

var Col = function (_Component) {
  _inherits(Col, _Component);

  function Col() {
    _classCallCheck(this, Col);

    return _possibleConstructorReturn(this, (Col.__proto__ || Object.getPrototypeOf(Col)).apply(this, arguments));
  }

  _createClass(Col, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          props = _objectWithoutProperties(_props, ['children', 'className']);

      var boolClassNames = [];

      BOOL_PROPS.forEach(function (name) {
        if (props[name]) {
          boolClassNames.push(BOOL_PROPS_TO_CLASS_NAMES[name]);
        }
        delete props[name];
      });

      var numberClassNames = [];

      NUMBER_PROPS.forEach(function (name) {
        var value = props[name];
        var func = NUMBER_PROPS_TO_CLASS_NAMES[name];

        if (value != null) {
          numberClassNames.push(func(value));
        }
        delete props[name];
      });

      var divClassName = (0, _classnames2.default)(className, 'rev-Col', boolClassNames, numberClassNames);

      return _react2.default.createElement(
        'div',
        _extends({}, props, { className: divClassName }),
        children
      );
    }
  }]);

  return Col;
}(_react.Component);

Col.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};
exports.default = Col;
