// harmonium/lib/DataGrid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _grid = require('./grid');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BOOL_PROPS_TO_CLASS_NAMES = {
  striped: 'rev-DataGrid--striped',
  horizontalScroll: 'rev-DataGrid--horizontalScroll',
  verticalScroll: 'rev-DataGrid--verticalScroll',
  stacked: 'rev-DataGrid--stacked',
  stackForSmall: 'rev-DataGrid--stackForSmall',
  stackForMedium: 'rev-DataGrid--stackForMedium',
  stackForLarge: 'rev-DataGrid--stackForLarge'
};
var BOOL_PROPS = Object.keys(BOOL_PROPS_TO_CLASS_NAMES);

var DataGrid = function (_React$Component) {
  _inherits(DataGrid, _React$Component);

  function DataGrid() {
    _classCallCheck(this, DataGrid);

    return _possibleConstructorReturn(this, (DataGrid.__proto__ || Object.getPrototypeOf(DataGrid)).apply(this, arguments));
  }

  _createClass(DataGrid, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          children = _props.children,
          props = _objectWithoutProperties(_props, ['className', 'children']);

      var propClassNames = [];

      BOOL_PROPS.forEach(function (name) {
        if (props[name]) {
          propClassNames.push(BOOL_PROPS_TO_CLASS_NAMES[name]);
        }
        delete props[name];
      });

      var newClassName = (0, _classnames2.default)(className, 'rev-DataGrid', propClassNames);

      return _react2.default.createElement(
        'div',
        _extends({}, props, { className: newClassName }),
        children
      );
    }
  }]);

  return DataGrid;
}(_react2.default.Component);

DataGrid.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};
exports.default = DataGrid;

var DataGridHeaderRow = function (_Component) {
  _inherits(DataGridHeaderRow, _Component);

  function DataGridHeaderRow() {
    _classCallCheck(this, DataGridHeaderRow);

    return _possibleConstructorReturn(this, (DataGridHeaderRow.__proto__ || Object.getPrototypeOf(DataGridHeaderRow)).apply(this, arguments));
  }

  _createClass(DataGridHeaderRow, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          className = _props2.className,
          children = _props2.children,
          props = _objectWithoutProperties(_props2, ['className', 'children']);

      return _react2.default.createElement(
        _grid.Row,
        _extends({}, props, { className: 'rev-DataGrid-headerRow ' + className }),
        children
      );
    }
  }]);

  return DataGridHeaderRow;
}(_react.Component);

DataGridHeaderRow.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.HeaderRow = DataGridHeaderRow;

var DataGridHeaderCol = function (_Component2) {
  _inherits(DataGridHeaderCol, _Component2);

  function DataGridHeaderCol() {
    _classCallCheck(this, DataGridHeaderCol);

    return _possibleConstructorReturn(this, (DataGridHeaderCol.__proto__ || Object.getPrototypeOf(DataGridHeaderCol)).apply(this, arguments));
  }

  _createClass(DataGridHeaderCol, [{
    key: 'render',
    value: function render() {
      var _props3 = this.props,
          className = _props3.className,
          children = _props3.children,
          props = _objectWithoutProperties(_props3, ['className', 'children']);

      return _react2.default.createElement(
        _grid.Col,
        _extends({}, props, { className: 'rev-DataGrid-headerCol ' + className }),
        children
      );
    }
  }]);

  return DataGridHeaderCol;
}(_react.Component);

DataGridHeaderCol.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.HeaderCol = DataGridHeaderCol;

var DataGridHeaderRowStacked = function (_Component3) {
  _inherits(DataGridHeaderRowStacked, _Component3);

  function DataGridHeaderRowStacked() {
    _classCallCheck(this, DataGridHeaderRowStacked);

    return _possibleConstructorReturn(this, (DataGridHeaderRowStacked.__proto__ || Object.getPrototypeOf(DataGridHeaderRowStacked)).apply(this, arguments));
  }

  _createClass(DataGridHeaderRowStacked, [{
    key: 'render',
    value: function render() {
      var _props4 = this.props,
          className = _props4.className,
          children = _props4.children,
          props = _objectWithoutProperties(_props4, ['className', 'children']);

      return _react2.default.createElement(
        _grid.Row,
        _extends({}, props, { className: 'rev-DataGrid-headerRow rev-DataGrid-headerRow--stacked ' + className }),
        children
      );
    }
  }]);

  return DataGridHeaderRowStacked;
}(_react.Component);

DataGridHeaderRowStacked.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.HeaderRowStacked = DataGridHeaderRowStacked;

var DataGridHeaderInline = function (_Component4) {
  _inherits(DataGridHeaderInline, _Component4);

  function DataGridHeaderInline() {
    _classCallCheck(this, DataGridHeaderInline);

    return _possibleConstructorReturn(this, (DataGridHeaderInline.__proto__ || Object.getPrototypeOf(DataGridHeaderInline)).apply(this, arguments));
  }

  _createClass(DataGridHeaderInline, [{
    key: 'render',
    value: function render() {
      var _props5 = this.props,
          className = _props5.className,
          children = _props5.children,
          props = _objectWithoutProperties(_props5, ['className', 'children']);

      return _react2.default.createElement(
        'span',
        _extends({}, props, { className: 'rev-DataGrid-header--inline ' + className }),
        children
      );
    }
  }]);

  return DataGridHeaderInline;
}(_react.Component);

DataGridHeaderInline.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.HeaderInline = DataGridHeaderInline;

var DataGridBody = function (_Component5) {
  _inherits(DataGridBody, _Component5);

  function DataGridBody() {
    _classCallCheck(this, DataGridBody);

    return _possibleConstructorReturn(this, (DataGridBody.__proto__ || Object.getPrototypeOf(DataGridBody)).apply(this, arguments));
  }

  _createClass(DataGridBody, [{
    key: 'render',
    value: function render() {
      var _props6 = this.props,
          className = _props6.className,
          children = _props6.children;


      return _react2.default.createElement(
        'div',
        { className: 'rev-DataGrid-body ' + className },
        children
      );
    }
  }]);

  return DataGridBody;
}(_react.Component);

DataGridBody.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.Body = DataGridBody;

var DataGridRow = function (_Component6) {
  _inherits(DataGridRow, _Component6);

  function DataGridRow() {
    _classCallCheck(this, DataGridRow);

    return _possibleConstructorReturn(this, (DataGridRow.__proto__ || Object.getPrototypeOf(DataGridRow)).apply(this, arguments));
  }

  _createClass(DataGridRow, [{
    key: 'render',
    value: function render() {
      var _props7 = this.props,
          className = _props7.className,
          children = _props7.children;


      return _react2.default.createElement(
        _grid.Row,
        { className: 'rev-DataGrid-row ' + className },
        children
      );
    }
  }]);

  return DataGridRow;
}(_react.Component);

DataGridRow.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.Row = DataGridRow;

var DataGridCol = function (_Component7) {
  _inherits(DataGridCol, _Component7);

  function DataGridCol() {
    _classCallCheck(this, DataGridCol);

    return _possibleConstructorReturn(this, (DataGridCol.__proto__ || Object.getPrototypeOf(DataGridCol)).apply(this, arguments));
  }

  _createClass(DataGridCol, [{
    key: 'render',
    value: function render() {
      var _props8 = this.props,
          className = _props8.className,
          children = _props8.children,
          props = _objectWithoutProperties(_props8, ['className', 'children']);

      return _react2.default.createElement(
        _grid.Col,
        _extends({}, props, { className: 'rev-DataGrid-col ' + className }),
        children
      );
    }
  }]);

  return DataGridCol;
}(_react.Component);

DataGridCol.propTypes = {
  className: _propTypes2.default.string,
  children: _propTypes2.default.node
};

DataGrid.Col = DataGridCol;
