// harmonium/lib/Input.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _lodash = require('lodash');

var _InputLabel = require('./InputLabel');

var _InputLabel2 = _interopRequireDefault(_InputLabel);

var _InputHelpText = require('./InputHelpText');

var _InputHelpText2 = _interopRequireDefault(_InputHelpText);

var _InputErrors = require('./InputErrors');

var _InputErrors2 = _interopRequireDefault(_InputErrors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Input = function (_Component) {
  _inherits(Input, _Component);

  function Input() {
    _classCallCheck(this, Input);

    return _possibleConstructorReturn(this, (Input.__proto__ || Object.getPrototypeOf(Input)).apply(this, arguments));
  }

  _createClass(Input, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          id = _props.id,
          className = _props.className,
          error = _props.error,
          inputRef = _props.inputRef,
          props = _objectWithoutProperties(_props, ['id', 'className', 'error', 'inputRef']);

      var inputClassName = (0, _classnames2.default)(className, 'rev-Input', {
        'is-invalid-input': !!error,
        'is-invalid': !!error
      });
      var inputProps = _extends({}, props, {
        className: inputClassName,
        ref: inputRef
      });

      return _react2.default.createElement('input', _extends({ id: id }, inputProps));
    }
  }]);

  return Input;
}(_react.Component);

Input.propTypes = {
  error: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.string]),
  inputRef: _propTypes2.default.func,
  id: _propTypes2.default.string,
  className: _propTypes2.default.string
};
Input.defaultProps = {
  type: 'text'
};
exports.default = Input;

var InputStack = function (_Component2) {
  _inherits(InputStack, _Component2);

  function InputStack() {
    _classCallCheck(this, InputStack);

    return _possibleConstructorReturn(this, (InputStack.__proto__ || Object.getPrototypeOf(InputStack)).apply(this, arguments));
  }

  _createClass(InputStack, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          error = _props2.error,
          help = _props2.help,
          className = _props2.className,
          label = _props2.label,
          props = _objectWithoutProperties(_props2, ['error', 'help', 'className', 'label']);

      var labelClassName = (0, _classnames2.default)(className, 'rev-InputStack');
      var inputId = (0, _lodash.uniqueId)('Input:');

      return _react2.default.createElement(
        _InputLabel2.default,
        { className: labelClassName, error: error, inputId: inputId },
        label,
        _react2.default.createElement(Input, _extends({}, props, { error: error, id: inputId })),
        _react2.default.createElement(
          _InputHelpText2.default,
          null,
          help
        ),
        _react2.default.createElement(
          _InputErrors2.default,
          null,
          error
        )
      );
    }
  }]);

  return InputStack;
}(_react.Component);

InputStack.propTypes = {
  error: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.string]),
  help: _propTypes2.default.string,
  label: _propTypes2.default.string,
  className: _propTypes2.default.string
};

Input.Stack = InputStack;
