// harmonium/lib/DatePicker/DateInputBlock.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Input = require('../Input');

var _Input2 = _interopRequireDefault(_Input);

var _createElementWithOverride = require('../Utilities/createElementWithOverride');

var _createElementWithOverride2 = _interopRequireDefault(_createElementWithOverride);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /** @jsx createElement */


/**
 * A component which contains the input(s) for a DatePicker. If the DatePicker
 * is a true type="date" input, and we aren't overriding the default format,
 * there will be one input here-- the input[type="date"]. If the browser has bad
 * type="date" support, or we chosen a custom date format, there will be two
 * inputs. One is a visible text input where the user types / calendar sets
 * local formatted date values (e.g. 03/12/2018). The other will be a hidden
 * input which carries the iso date value that a true date field would output.
 * Only the hidden field in this case has a name= attribute, and so it is the
 * only value submitted. This allows the server to expect the same format from
 * the client, whether date inputs are well supported on the client or not.
 * @param {object} props - the props of the DateInputBlock
 */
var DateInputBlock = function (_Component) {
  _inherits(DateInputBlock, _Component);

  function DateInputBlock() {
    _classCallCheck(this, DateInputBlock);

    return _possibleConstructorReturn(this, (DateInputBlock.__proto__ || Object.getPrototypeOf(DateInputBlock)).apply(this, arguments));
  }

  _createClass(DateInputBlock, [{
    key: 'render',


    /* eslint complexity: [2, 6] */
    value: function render() {
      var _props = this.props,
          error = _props.error,
          className = _props.className,
          goodDateInput = _props.goodDateInput,
          generation = _props.generation,
          overrides = _props.overrides,
          dateFormat = _props.dateFormat,
          isoValue = _props.isoValue,
          formattedValue = _props.formattedValue,
          name = _props.name,
          props = _objectWithoutProperties(_props, ['error', 'className', 'goodDateInput', 'generation', 'overrides', 'dateFormat', 'isoValue', 'formattedValue', 'name']);

      var createElement = _createElementWithOverride2.default.bind(this, overrides);
      var inputClassName = (0, _classnames2.default)(className, 'rev-DatePicker-input', {
        'is-invalid-input': !!error,
        'is-invalid': !!error
      });

      return createElement(
        'div',
        null,
        createElement(_Input2.default, _extends({}, props, {
          className: inputClassName,
          type: goodDateInput ? 'date' : 'text',
          name: goodDateInput ? name : null,
          defaultValue: formattedValue
          /*         have a placeholder to avoid empty box on Firefox  */
          , placeholder: dateFormat ? dateFormat : 'mm/dd/yyyy'
        })),
        goodDateInput ? null : createElement(_Input2.default, {
          type: 'hidden',
          name: name,
          key: generation + ':trueInput',
          value: isoValue || '',
          readOnly: true
        })
      );
    }
  }]);

  return DateInputBlock;
}(_react.Component);

DateInputBlock.propTypes = {
  error: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.string]),
  goodDateInput: _propTypes2.default.bool,
  generation: _propTypes2.default.number,
  overrides: _propTypes2.default.object,
  dateFormat: _propTypes2.default.string,
  isoValue: _propTypes2.default.string,
  formattedValue: _propTypes2.default.string,
  name: _propTypes2.default.string,
  className: _propTypes2.default.string };
exports.default = DateInputBlock;
